File: example/language/hello.rb
#!/usr/bin/env ruby

# Ruby示例文件
# 展示Ruby的各种语法特性

require 'date'
require 'time'

# 全局常量
PI = 3.14159
APP_NAME = "Ruby Example"

# 模块定义
module MathUtils
  def self.square(x)
    x * x
  end

  def self.cube(x)
    x * x * x
  end
end

# 异常类
class ValidationError < StandardError
  def initialize(message = "Validation failed")
    super(message)
  end
end

# 类定义
class Person
  # 类变量
  @@person_count = 0

  # 属性访问器
  attr_accessor :name, :age, :email
  attr_reader :id

  # 初始化方法
  def initialize(name, age, email = nil)
    @id = generate_id
    @name = name
    @age = age
    @email = email
    @@person_count += 1
  end

  # 类方法
  def self.total_count
    @@person_count
  end

  def self.create_from_hash(hash)
    new(hash[:name], hash[:age], hash[:email])
  end

  # 实例方法
  def greet
    "Hello, my name is #{@name} and I'm #{@age} years old"
  end

  def adult?
    @age >= 18
  end

  def birthday
    @age += 1
    "Happy birthday! You're now #{@age} years old"
  end

  def to_s
    "#{@name} (#{@age})"
  end

  private

  def generate_id
    rand(10000..99999)
  end
end

# 继承
class Employee < Person
  attr_accessor :position, :salary

  def initialize(name, age, position, salary, email = nil)
    super(name, age, email)
    @position = position
    @salary = salary
  end

  def promote(new_position, new_salary)
    @position = new_position
    @salary = new_salary
    "#{@name} has been promoted to #{@position}"
  end

  def to_s
    super + " - #{@position} ($#{@salary})"
  end
end

# 模块包含（Mixins）
module Debuggable
  def debug_info
    "Object ID: #{object_id}, Class: #{self.class}"
  end
end

class Calculator
  include Debuggable

  def initialize
    @result = 0
  end

  def add(number)
    @result += number
  end

  def subtract(number)
    @result -= number
  end

  def multiply(number)
    @result *= number
  end

  def divide(number)
    raise ZeroDivisionError, "Cannot divide by zero" if number.zero?
    @result /= number.to_f
  end

  def result
    @result
  end

  def reset
    @result = 0
  end
end

# 代码块和迭代器
def fibonacci(n)
  return [] if n <= 0
  return [0] if n == 1

  sequence = [0, 1]
  (2...n).each do |i|
    sequence << sequence[i-1] + sequence[i-2]
  end
  sequence
end

def process_array(array, &block)
  array.map(&block)
end

# DSL 示例
class Config
  def initialize(&block)
    @settings = {}
    instance_eval(&block) if block_given?
  end

  def set(key, value)
    @settings[key] = value
  end

  def get(key)
    @settings[key]
  end

  def to_hash
    @settings.dup
  end
end

# 正则表达式
class StringValidator
  EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
  PHONE_REGEX = /\A\d{3}-\d{3}-\d{4}\z/

  def self.valid_email?(email)
    !!(email =~ EMAIL_REGEX)
  end

  def self.valid_phone?(phone)
    !!(phone =~ PHONE_REGEX)
  end
end

# 主程序
def main
  puts "=== Ruby Example ==="

  # 基本数据类型
  puts "\n--- Basic Data Types ---"
  string = "Hello World"
  number = 42
  float = 3.14159
  boolean = true
  array = [1, 2, 3, 4, 5]
  hash = {name: "Ruby", version: "3.0", awesome: true}

  puts "String: #{string}"
  puts "Number: #{number}"
  puts "Float: #{float}"
  puts "Boolean: #{boolean}"
  puts "Array: #{array.inspect}"
  puts "Hash: #{hash.inspect}"

  # 字符串插值和操作
  puts "\n--- String Operations ---"
  name = "Alice"
  age = 30
  puts "My name is #{name} and I am #{age} years old"
  puts "String length: #{string.length}"
  puts "Uppercase: #{string.upcase}"
  puts "Reversed: #{string.reverse}"

  # 数组操作
  puts "\n--- Array Operations ---"
  numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
  puts "Original: #{numbers.inspect}"
  puts "Even numbers: #{numbers.select(&:even?).inspect}"
  puts "Sum: #{numbers.sum}"
  puts "Average: #{numbers.sum / numbers.length.to_f}"

  # 哈希操作
  puts "\n--- Hash Operations ---"
  person_data = {name: "Bob", age: 25, city: "New York"}
  puts "Person: #{person_data.inspect}"
  puts "Keys: #{person_data.keys.inspect}"
  puts "Values: #{person_data.values.inspect}"

  # 条件语句
  puts "\n--- Conditional Statements ---"
  score = 85

  if score >= 90
    grade = "A"
  elsif score >= 80
    grade = "B"
  elsif score >= 70
    grade = "C"
  else
    grade = "F"
  end

  puts "Score: #{score}, Grade: #{grade}"

  # 循环
  puts "\n--- Loops ---"
  puts "For loop:"
  for i in 1..5
    puts "  Iteration #{i}"
  end

  puts "Each loop:"
  ["apple", "banana", "cherry"].each do |fruit|
    puts "  I like #{fruit}"
  end

  puts "Times loop:"
  3.times do |i|
    puts "  Count: #{i}"
  end

  # 类和对象
  puts "\n--- Classes and Objects ---"
  person1 = Person.new("Charlie", 28, "charlie@example.com")
  person2 = Person.new("Diana", 32)

  puts person1.greet
  puts person2.greet
  puts "Total persons created: #{Person.total_count}"

  employee = Employee.new("Eve", 35, "Developer", 75000)
  puts employee.to_s
  puts employee.promote("Senior Developer", 95000)
  puts employee.to_s

  # 异常处理
  puts "\n--- Exception Handling ---"
  begin
    result = 10 / 0
  rescue ZeroDivisionError => e
    puts "Caught division by zero: #{e.message}"
  rescue => e
    puts "Caught general error: #{e.message}"
  ensure
    puts "This always executes"
  end

  # 计算器使用
  puts "\n--- Calculator Usage ---"
  calc = Calculator.new
  calc.add(10)
  calc.multiply(2)
  calc.subtract(3)
  puts "Result: #{calc.result}"
  puts calc.debug_info

  # 斐波那契数列
  puts "\n--- Fibonacci Sequence ---"
  fib_sequence = fibonacci(10)
  puts "First 10 Fibonacci numbers: #{fib_sequence.inspect}"

  # 代码块处理
  puts "\n--- Block Processing ---"
  numbers = [1, 2, 3, 4, 5]
  doubled = process_array(numbers) { |x| x * 2 }
  puts "Original: #{numbers.inspect}"
  puts "Doubled: #{doubled.inspect}"

  # DSL使用
  puts "\n--- DSL Usage ---"
  config = Config.new do
    set :database_url, "sqlite:///app.db"
    set :debug_mode, true
    set :max_connections, 100
  end

  puts "Config: #{config.to_hash.inspect}"

  # 正则表达式验证
  puts "\n--- Regular Expression Validation ---"
  emails = ["valid@example.com", "invalid-email", "another@valid.com"]
  phones = ["123-456-7890", "1234567890", "123-45-6789"]

  puts "Valid emails:"
  emails.each { |email| puts "  #{email}: #{StringValidator.valid_email?(email)}" }

  puts "Valid phones:"
  phones.each { |phone| puts "  #{phone}: #{StringValidator.valid_phone?(phone)}" }

  # 日期和时间
  puts "\n--- Date and Time ---"
  now = Time.now
  today = Date.today

  puts "Current time: #{now}"
  puts "Current date: #{today}"
  puts "Day of week: #{today.strftime('%A')}"

  # 范围和区间
  puts "\n--- Ranges ---"
  range = (1..10)
  puts "Range 1..10 includes 5: #{range.include?(5)}"
  puts "Range 1..10 includes 15: #{range.include?(15)}"

  char_range = ('a'..'z')
  puts "First 5 letters: #{char_range.first(5).inspect}"

  # 符号（Symbols）
  puts "\n--- Symbols ---"
  symbol1 = :hello
  symbol2 = :hello
  string1 = "hello"
  string2 = "hello"

  puts "symbol1 == symbol2: #{symbol1 == symbol2}"
  puts "string1 == string2: #{string1 == string2}"
  puts "symbol1 == string1: #{symbol1 == string1}"
  puts "symbol1.object_id == symbol2.object_id: #{symbol1.object_id == symbol2.object_id}"
  puts "string1.object_id == string2.object_id: #{string1.object_id == string2.object_id}"

  puts "\n=== Ruby Example Completed ==="
end

# 脚本入口
if __FILE__ == $0
  main
end

File: example/language/hello.swift
// Swift 示例文件
// 展示 Swift 的各种语法特性

import Foundation

// 全局常量和变量
let appName = "Swift Example"
let version = "1.0.0"
var globalCounter = 0

// 枚举
enum Status {
    case pending
    case approved
    case rejected

    var description: String {
        switch self {
        case .pending:
            return "Pending Review"
        case .approved:
            return "Approved"
        case .rejected:
            return "Rejected"
        }
    }
}

enum Result<T, Error> {
    case success(T)
    case failure(Error)
}

// 协议（Protocol）
protocol Describable {
    var description: String { get }
}

protocol Logger {
    func log(_ message: String)
    func error(_ message: String)
}

// 结构体
struct Point: Describable {
    var x: Double
    var y: Double

    var description: String {
        return "Point(x: \(x), y: \(y))"
    }

    func distance(to other: Point) -> Double {
        let dx = x - other.x
        let dy = y - other.y
        return sqrt(dx * dx + dy * dy)
    }
}

struct Size {
    var width: Double
    var height: Double
}

// 类
class Person: Describable {
    // 属性
    let id: UUID
    var name: String
    var age: Int
    var email: String?

    // 计算属性
    var isAdult: Bool {
        return age >= 18
    }

    // 懒加载属性
    lazy var displayName: String = {
        return "\(name) (\(age))"
    }()

    // 初始化器
    init(name: String, age: Int, email: String? = nil) {
        self.id = UUID()
        self.name = name
        self.age = age
        self.email = email
    }

    // 便利初始化器
    convenience init(name: String) {
        self.init(name: name, age: 0, email: nil)
    }

    // 方法
    func greet() -> String {
        return "Hello, my name is \(name)"
    }

    func celebrateBirthday() -> String {
        age += 1
        return "Happy birthday! You're now \(age) years old"
    }

    // 协议实现
    var description: String {
        return "Person: \(name), Age: \(age)"
    }
}

// 继承
class Employee: Person {
    var position: String
    var salary: Double

    init(name: String, age: Int, position: String, salary: Double, email: String? = nil) {
        self.position = position
        self.salary = salary
        super.init(name: name, age: age, email: email)
    }

    override func greet() -> String {
        return "Hello, I'm \(name), a \(position)"
    }

    func getAnnualSalary() -> Double {
        return salary * 12
    }
}

// 扩展（Extensions）
extension Double {
    var squared: Double {
        return self * self
    }

    var cubed: Double {
        return self * self * self
    }
}

extension Array where Element: Describable {
    func describeAll() -> String {
        return self.map { $0.description }.joined(separator: "\n")
    }
}

// 泛型
struct Stack<Element> {
    private var elements: [Element] = []

    mutating func push(_ element: Element) {
        elements.append(element)
    }

    mutating func pop() -> Element? {
        return elements.popLast()
    }

    func peek() -> Element? {
        return elements.last
    }

    var isEmpty: Bool {
        return elements.isEmpty
    }

    var count: Int {
        return elements.count
    }
}

// 错误处理
enum CalculatorError: Error {
    case divisionByZero
    case invalidInput(String)
}

class Calculator {
    private var result: Double = 0

    func add(_ number: Double) {
        result += number
    }

    func subtract(_ number: Double) {
        result -= number
    }

    func multiply(_ number: Double) {
        result *= number
    }

    func divide(_ number: Double) throws {
        guard number != 0 else {
            throw CalculatorError.divisionByZero
        }
        result /= number
    }

    func getResult() -> Double {
        return result
    }

    func reset() {
        result = 0
    }
}

// 闭包和函数
func fibonacci(_ n: Int) -> [Int] {
    guard n > 0 else { return [] }
    if n == 1 { return [0] }

    var sequence = [0, 1]
    for i in 2..<n {
        sequence.append(sequence[i-1] + sequence[i-2])
    }
    return sequence
}

func processArray(_ array: [Int], using transform: (Int) -> Int) -> [Int] {
    return array.map(transform)
}

func createMultiplier(_ factor: Int) -> (Int) -> Int {
    return { number in
        return number * factor
    }
}

// 操作符重载
struct Vector2D {
    var x: Double
    var y: Double

    static func + (left: Vector2D, right: Vector2D) -> Vector2D {
        return Vector2D(x: left.x + right.x, y: left.y + right.y)
    }

    static func * (vector: Vector2D, scalar: Double) -> Vector2D {
        return Vector2D(x: vector.x * scalar, y: vector.y * scalar)
    }
}

// 模式匹配和guard语句
func describeNumber(_ number: Int) -> String {
    switch number {
    case 0:
        return "zero"
    case 1...9:
        return "single digit"
    case 10...99:
        return "double digits"
    case let x where x % 2 == 0:
        return "even number"
    default:
        return "other number"
    }
}

// 主函数
func main() {
    print("=== Swift Example ===")

    // 基本数据类型
    print("\n--- Basic Data Types ---")
    let string: String = "Hello World"
    let integer: Int = 42
    let float: Double = 3.14159
    let boolean: Bool = true
    let array: [Int] = [1, 2, 3, 4, 5]
    let dictionary: [String: Any] = ["name": "Swift", "version": "5.7", "awesome": true]

    print("String: \(string)")
    print("Integer: \(integer)")
    print("Float: \(float)")
    print("Boolean: \(boolean)")
    print("Array: \(array)")
    print("Dictionary: \(dictionary)")

    // 可选类型
    print("\n--- Optionals ---")
    let optionalString: String? = "Optional value"
    let nilString: String? = nil

    if let unwrapped = optionalString {
        print("Optional contains: \(unwrapped)")
    }

    let defaultValue = nilString ?? "Default value"
    print("Nil coalescing result: \(defaultValue)")

    // 字符串插值和操作
    print("\n--- String Operations ---")
    let name = "Alice"
    let age = 30
    let message = "My name is \(name) and I am \(age) years old"
    print(message)
    print("String length: \(string.count)")
    print("Uppercase: \(string.uppercased())")
    print("Reversed: \(String(string.reversed()))")

    // 数组操作
    print("\n--- Array Operations ---")
    let numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    let evenNumbers = numbers.filter { $0 % 2 == 0 }
    let squaredNumbers = numbers.map { $0 * $0 }
    let sum = numbers.reduce(0, +)

    print("Original: \(numbers)")
    print("Even numbers: \(evenNumbers)")
    print("Squared: \(squaredNumbers)")
    print("Sum: \(sum)")

    // 条件语句
    print("\n--- Conditional Statements ---")
    let score = 85

    let grade: String
    if score >= 90 {
        grade = "A"
    } else if score >= 80 {
        grade = "B"
    } else if score >= 70 {
        grade = "C"
    } else {
        grade = "F"
    }

    print("Score: \(score), Grade: \(grade)")

    // 循环
    print("\n--- Loops ---")
    print("For-in loop:")
    for i in 1...5 {
        print("  Iteration \(i)")
    }

    print("For-in with array:")
    for fruit in ["apple", "banana", "cherry"] {
        print("  I like \(fruit)")
    }

    print("While loop:")
    var counter = 1
    while counter <= 3 {
        print("  Count: \(counter)")
        counter += 1
    }

    // 类和对象
    print("\n--- Classes and Objects ---")
    let person = Person(name: "Charlie", age: 28, email: "charlie@example.com")
    let employee = Employee(name: "Diana", age: 35, position: "Developer", salary: 75000)

    print(person.greet())
    print(employee.greet())
    print("Employee annual salary: \(employee.getAnnualSalary())")

    // 结构体使用
    let point1 = Point(x: 0, y: 0)
    let point2 = Point(x: 3, y: 4)
    print("Point 1: \(point1)")
    print("Point 2: \(point2)")
    print("Distance between points: \(point1.distance(to: point2))")

    // 枚举使用
    let status = Status.approved
    print("Status: \(status.description)")

    // 泛型栈
    var intStack = Stack<Int>()
    intStack.push(1)
    intStack.push(2)
    intStack.push(3)
    print("Stack count: \(intStack.count)")
    print("Popped: \(intStack.pop() ?? 0)")
    print("Stack count after pop: \(intStack.count)")

    // 计算器使用和错误处理
    print("\n--- Calculator and Error Handling ---")
    let calc = Calculator()
    calc.add(10)
    calc.multiply(2)

    do {
        try calc.divide(0)
    } catch CalculatorError.divisionByZero {
        print("Caught division by zero error")
    } catch CalculatorError.invalidInput(let message) {
        print("Invalid input: \(message)")
    } catch {
        print("Unknown error: \(error)")
    }

    print("Calculator result: \(calc.getResult())")

    // 函数调用
    print("\n--- Function Calls ---")
    let fibSequence = fibonacci(10)
    print("Fibonacci sequence: \(fibSequence)")

    let doubled = processArray([1, 2, 3, 4, 5]) { $0 * 2 }
    print("Doubled array: \(doubled)")

    let multiplier = createMultiplier(3)
    print("5 * 3 = \(multiplier(5))")

    // 扩展使用
    print("\n--- Extensions ---")
    let value: Double = 5.0
    print("5 squared: \(value.squared)")
    print("5 cubed: \(value.cubed)")

    let describableObjects: [Describable] = [point1, person]
    print("All objects:\n\(describableObjects.describeAll())")

    // 操作符重载
    print("\n--- Operator Overloading ---")
    let vector1 = Vector2D(x: 1, y: 2)
    let vector2 = Vector2D(x: 3, y: 4)
    let sumVector = vector1 + vector2
    let scaledVector = vector1 * 2.0

    print("Vector 1: (\(vector1.x), \(vector1.y))")
    print("Vector 2: (\(vector2.x), \(vector2.y))")
    print("Sum: (\(sumVector.x), \(sumVector.y))")
    print("Scaled: (\(scaledVector.x), \(scaledVector.y))")

    // 模式匹配
    print("\n--- Pattern Matching ---")
    for num in [0, 5, 15, 42, 100] {
        print("\(num) is \(describeNumber(num))")
    }

    // 日期和时间
    print("\n--- Date and Time ---")
    let now = Date()
    let formatter = DateFormatter()
    formatter.dateStyle = .medium
    formatter.timeStyle = .medium

    print("Current date and time: \(formatter.string(from: now))")

    // 文件操作（基本示例）
    print("\n--- File Operations ---")
    let tempDir = NSTemporaryDirectory()
    let tempFile = tempDir + "swift_example.txt"

    do {
        let content = "This is a test file created by Swift example.\n"
        try content.write(toFile: tempFile, atomically: true, encoding: .utf8)

        let readContent = try String(contentsOfFile: tempFile, encoding: .utf8)
        print("File content: \(readContent.trimmingCharacters(in: .whitespacesAndNewlines))")

        // 清理
        try FileManager.default.removeItem(atPath: tempFile)
        print("Temporary file cleaned up")

    } catch {
        print("File operation error: \(error)")
    }

    print("\n=== Swift Example Completed ===")
}

// 程序入口
main()
